aten/src/ATen/BatchedFallback.cpp

```diff
@@ -0,0 +1,137 @@
+}
+
+static bool areAnyArgumentsTensorList(const FunctionSchema& schema) {
+  return std::any_of(
+      schema.arguments().begin(),
+      schema.arguments().end(),
+      [] (const Argument& arg) { return arg.type()->isSubtypeOf(ListType::ofTensors()); });
+}
+
+void batchedTensorForLoopFallback(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
+  const auto& schema = op.schema();
+  const auto num_returns = schema.returns().size();
+  TORCH_CHECK(!schema.is_mutable() && !schema.hasAnyAliasInfo(),
+              "Batching rule not implemented for ", schema, "; ",
+              "the fallback path doesn't work on in-place or view ops.");
+  TORCH_CHECK(areAllReturnsTensors(schema) && !areAnyArgumentsTensorList(schema),
+              "Batching rule not implemented for ", schema, ". ",
+              "We could not generate a fallback.");
+  TORCH_CHECK(num_returns == 1,
+              "Batching rule not implemented for ", schema, ". ",
+              "We do not yet support operations with multiple returns.");
+  TORCH_WARN("Batching rule not implemented for ", schema, " falling back "
+             "to slow (for loop and stack) implementation");
+
+  const auto num_arguments = schema.arguments().size();
+  const auto arguments = torch::jit::last(stack, num_arguments);
+
+  // Figure out which arguments are BatchedTensor. Save them to a vector.
+  // For each BatchedTensor, also record what position of `arguments` they came from.
+  SmallVector<Tensor,kVmapTransformStaticInputSize> batched_tensor_inputs;
+  VmapDimVector batched_tensor_inputs_position;
+  for (int64_t idx = 0; idx < arguments.size(); ++idx) {
+    const auto& ivalue = arguments[idx];
+    if (!ivalue.isTensor()) {
+      continue;
+    }
+    const auto& tensor = ivalue.toTensor();
+    if (!tensor.defined()) {
+      continue;
+    }
+    const auto* batched = maybeGetBatched(tensor);
+    if (!batched) {
+      continue;
+    }
+    batched_tensor_inputs.push_back(tensor);
+    batched_tensor_inputs_position.push_back(idx);
+  }
+  TORCH_INTERNAL_ASSERT(batched_tensor_inputs.size() > 0);
+
+  // MultiBatchVmapTransform the BatchedTensor arguments. This returns
+  // VmapPhysicalViews that contain all of the batch dimensions.
+  const auto input_physical_views = MultiBatchVmapTransform::logicalToPhysical(
+      batched_tensor_inputs);
+
+  // Compute the total number of batches
+  auto num_batch_dims = input_physical_views.front().numBatchDims();
+  auto some_sizes = input_physical_views.front().tensor().sizes();
+  auto batch_sizes = ArrayRef<int64_t>(some_sizes.begin(), some_sizes.begin() + num_batch_dims);
+  auto num_batches = std::accumulate(
+      batch_sizes.begin(),
+      batch_sizes.end(),
+      1,
+      std::multiplies<int64_t>());
+
+  // Strategy: For each batch, we are going to push slices (where applicable)
+  // of the arguments onto `working_stack`, call `op`, and store the result in
+  // `output_shards`.
+  torch::jit::Stack working_stack;
```

Comment:
@ezyang I wanted to take this a step further and reuse the input stack. However, replacing this line with `auto& working_stack = *stack` causes some difficult-to-debug error messages like "RuntimeError: isTensor() INTERNAL ASSERT FAILED at "../aten/src/ATen/core/ivalue_inl.h":117, please report a bug to PyTorch. Expected Tensor but got None", "RuntimeError: IValue is not a Scalar". Perhaps there is existing code somewhere that assumes the stack must not contain additional arguments unrelated to the current operation?

Reply:
Can you get a full backtrace (and state of the stack) from gdb here? This sounds unexpected.

Reply:
I figured out the problem, it was a bug in my code. I've changed the code to just use the input stack.

If you're interested in the bug: somewhere above the code uses `torch::jit::last` to get an `ArrayRef<IValue>` of the inputs on the stack. But it turns out that when you modify the stack by pushing elements to it, the original `ArrayRef<IValue>` becomes garbage because the underlying `std::vector` that backs the stack is allowed to reallocate memory (and it does). My code was reading garbage values and that is how the errors happened.